src/server/game/Movement/MovementGenerators/PathGenerator.cpp

```diff
@@ -211,20 +211,21 @@ void PathGenerator::BuildPolyPath(G3D::Vector3 const& startPos, G3D::Vector3 con
     {
         bool buildShotrcut = false;
 
-        bool isUnderWaterStart = _source->GetMap()->IsUnderWater(_source->GetPhaseMask(), startPos.x, startPos.y, startPos.z, _source->GetCollisionHeight());
-        bool isUnderWaterEnd = _source->GetMap()->IsUnderWater(_source->GetPhaseMask(), endPos.x, endPos.y, endPos.z, _source->GetCollisionHeight());
-        bool isFarUnderWater = startFarFromPoly ? isUnderWaterStart : isUnderWaterEnd;
+        bool isWaterStart = _source->GetMap()->IsInWater(_source->GetPhaseMask(), startPos.x, startPos.y, startPos.z, _source->GetCollisionHeight());
```

Comment:
We can't use "IsInWater" here, this shortcut is made for those creatures which are fully under water (where the shortcut is possible), but when you are on the shore (just with the feet in water) the path **must not** use a shortcut otherwise they can have issues with textures.

What are you trying to solve here?

Reply:
I checked on shore and IsInWater returns false.
It's implementation is checking if the unit is swimming, so this should be OK.

The issue happens when you're floating on the surface of the water and attempt to charge a creature underwater, or the other way around. Also if both units are floating on the surface, path generation fails as well.

Reply:
@Yehonal what do you think? maybe it's a vmap/mmap related issue, i've seen several reports that TC is handling these things better.

Reply:
@Yehonal what do you think of this? I have checked several water positions, such as on the shore, where only feet are in water, and behavior is OK because we're checking if the unit is swimming.

Reply:
TrinityCore does not handle the long paths that contain both underwater and on-ground sections. For instance, if a creature chases you for a long path, once it has to go back to its own original position, the creature might swim to the shore, then run on the ground, then swim into a lake, then run again on the ground. This code above actually checks if the paths are composed of a mix of these 2 by correctly using shortcuts when it's in water and normal paths when the creature can walk.

Without that code (the trinitycore case) the creature would literally fly above the water in most of the cases and it continues to fly on the shore. TC applied some "tricks" with a combination of MMAPS + creature flags in order to allow the creatures not to do crazy stuff underwater. This code there is very generic instead and allows all creatures that can swim to move flawlessly.

Probably, we need to investigate a bit more on the **charge** issue when it's in water, but I'd not change this code if it breaks the behaviour I've just mentioned

Reply:
@Yehonal i just did a few tests pulling murlocs on the shore down to the water, and the path seems OK, they walk to the water, then swim towards me. When returning it does the same, swim towards the shore, then walking to it's original position.

Also tried pulling one under water, make it travel through an island again to the water on the other side, and then reset them, and also the path generated seems ok.

Is this the case you were talking about?
Is there any other case that should be checked?

Reply:
I mean, i see some weird walking in water, but it seems to be the same on master branch.
With this changes: https://youtu.be/WB9sjc27TDk
master: https://youtu.be/Uh7AFwUMfHU

It does seem pretty similar, with both I can see some degree of weird movement with creatures walking in the water instead of swimming, but they do climb the island accordingly.

Reply:
@Yehonal I can revert this particular change if you want, I have an incoming PR to update our maps to TC versions and it should fix this issues and some blink related ones as well.